pkg/controller/stacks/install/installjob_test.go

```diff
@@ -1486,21 +1490,100 @@ func TestCreateJobOutputObject(t *testing.T) {
+				client: func() client.Client {
+					crd := crd(withCRDGroupKind("samples.upbound.io", "Mytype"))
+					fc := fake.NewFakeClient(&crd)
+					return &test.MockClient{
+						MockCreate: fc.Create,
+						MockGet:    test.NewMockGetFn(errBoom),
+						MockUpdate: test.NewMockUpdateFn(errors.New("trap")),
+					}
+				}(),
+				log: logging.NewNopLogger(),
+			},
+			stackInstaller: resource(),
+			job:            job(),
+			obj:            unstructuredObj(crdRaw),
+			want: want{
+				err: errors.Wrap(errors.Wrap(errBoom, "failed to fetch existing crd"), "can not replace existing CRD"),
+				obj: nil,
+			},
+		},
+		{
+			name: "FailedCRDBeingDeleted",
+			jobCompleter: &stackInstallJobCompleter{
+				client: func() client.Client {
+					crd := crd(withCRDGroupKind("samples.upbound.io", "Mytype"), withCRDDeletionTimestamp(time.Date(2020, 04, 01, 0, 0, 0, 0, time.UTC)))
+					return fake.NewFakeClient(&crd)
+				}(),
+				log: logging.NewNopLogger(),
+			},
+			stackInstaller: resource(),
+			job:            job(),
+			obj:            unstructuredObj(crdRaw),
+			want: want{
+				err: errors.Wrap(errors.New("existing crd has not been deleted"), "can not replace existing CRD"),
+				obj: nil,
+			},
+		},
+		{
+			name: "FailedIncompatibleCRDExists",
+			jobCompleter: &stackInstallJobCompleter{
+				client: func() client.Client {
+					crd := crd(withCRDGroupKind("samples.upbound.io", "Mytype"), withCRDVersion("existing"))
+					return fake.NewFakeClient(&crd)
+				}(),
+				log: logging.NewNopLogger(),
+			},
+			stackInstaller: resource(),
+			job:            job(),
+			obj:            unstructuredObj(crdRaw),
+			want: want{
+				err: errors.Wrap(errors.New("replacement crd removes crd versions"), "can not replace existing CRD"),
+				obj: nil,
+			},
+		},
+		{
+			name: "SuccessUpdatingCRD",
+			jobCompleter: &stackInstallJobCompleter{
+				client: func() client.Client {
+					crd := crd(withCRDGroupKind("samples.upbound.io", "Mytype"), withCRDLabels(map[string]string{"foo": "bar"}))
+					return fake.NewFakeClient(&crd)
+				}(),
+				log: logging.NewNopLogger(),
+			},
+			stackInstaller: resource(),
+			job:            job(),
+			obj:            unstructuredObj(crdRaw, unstructuredAsCRD(withCRDVersion("new"))),
+			want: want{
+				err: nil,
+				obj: unstructuredObj(crdRaw, unstructuredAsCRD(withCRDVersion("new"), withCRDLabels(map[string]string{"foo": "bar"}))),
+			},
+		},
 	}
 
 	ctx := context.Background()
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			g := NewGomegaWithT(t)
+
 			gotErr := tt.jobCompleter.createJobOutputObject(ctx, tt.obj, tt.stackInstaller, tt.job)
 
 			if diff := cmp.Diff(tt.want.err, gotErr, test.EquateErrors()); diff != "" {
 				t.Errorf("createJobOutputObject(): -want error, +got error:\n%s", diff)
 			}
 
-			if diff := cmp.Diff(tt.want.obj, tt.obj, test.EquateConditions()); diff != "" {
-				t.Errorf("createJobOutputObject(): -want obj, +got obj:\n%v", diff)
```

Comment:
This test did not verify that the object that went into `createJobOutputObject` was the object that came out of `createJobOutputObject`.

The existing tests were told to expect that what went into the function is what went into the function (which would verify no mutations).

The change in testing here fetches the created resource and compares the expected object against that, which is more meaningful imho.